modules/core/04-channel/keeper/upgrade.go

```diff
@@ -95,6 +96,188 @@ func (k Keeper) WriteUpgradeInitChannel(
+			"connection state is not OPEN (got %s)", connectiontypes.State(connectionEnd.GetState()).String(),
+		)
+	}
+
+	if connectionEnd.GetCounterparty().GetConnectionID() != counterpartyChannel.ConnectionHops[0] {
+		return 0, "", errorsmod.Wrapf(connectiontypes.ErrInvalidConnection, "unexpected counterparty channel connection hops, expected %s but got %s", connectionEnd.GetCounterparty().GetConnectionID(), counterpartyChannel.ConnectionHops[0])
+	}
+
+	if err := k.connectionKeeper.VerifyChannelState(ctx, connectionEnd, proofHeight, proofChannel, proposedUpgradeChannel.Counterparty.PortId,
+		proposedUpgradeChannel.Counterparty.ChannelId, counterpartyChannel); err != nil {
+		return 0, "", err
+	}
+
+	upgradeTimeout := types.UpgradeTimeout{TimeoutHeight: timeoutHeight, TimeoutTimestamp: timeoutTimestamp}
+	if err := k.connectionKeeper.VerifyChannelUpgradeTimeout(ctx, connectionEnd, proofHeight, proofUpgradeTimeout, proposedUpgradeChannel.Counterparty.PortId,
+		proposedUpgradeChannel.Counterparty.ChannelId, upgradeTimeout); err != nil {
+		return 0, "", err
+	}
+
+	if err := k.connectionKeeper.VerifyChannelUpgradeSequence(ctx, connectionEnd, proofHeight, proofUpgradeSequence, proposedUpgradeChannel.Counterparty.PortId,
+		proposedUpgradeChannel.Counterparty.ChannelId, counterpartyUpgradeSequence); err != nil {
+		return 0, "", err
+	}
+
+	// check if upgrade timed out by comparing it with the latest height of the chain
+	selfHeight := clienttypes.GetSelfHeight(ctx)
+	if !timeoutHeight.IsZero() && selfHeight.GTE(timeoutHeight) {
+		return 0, "", k.RestoreChannel(ctx, portID, channelID, upgradeSequence, types.ErrUpgradeTimeout)
+	}
+
+	// check if upgrade timed out by comparing it with the latest timestamp of the chain
+	if timeoutTimestamp != 0 && uint64(ctx.BlockTime().UnixNano()) >= timeoutTimestamp {
+		return 0, "", k.RestoreChannel(ctx, portID, channelID, upgradeSequence, types.ErrUpgradeTimeout)
+	}
+
+	switch channel.State {
+	case types.OPEN:
+		upgradeSequence = uint64(0)
+		if seq, found := k.GetUpgradeSequence(ctx, portID, channelID); found {
+			upgradeSequence = seq
+		}
+
+		// if the counterparty upgrade sequence is ahead then fast forward so both channel ends are using the same sequence for the current upgrade
+		if counterpartyUpgradeSequence > upgradeSequence {
+			upgradeSequence = counterpartyUpgradeSequence
+			k.SetUpgradeSequence(ctx, portID, channelID, upgradeSequence)
+		} else {
+			errorReceipt := types.NewErrorReceipt(upgradeSequence, errorsmod.Wrapf(types.ErrInvalidUpgradeSequence, "upgrade sequence %d was not smaller than the counter party chain upgrade sequence %d", upgradeSequence, counterpartyUpgradeSequence))
+			// the upgrade sequence is incremented so both sides start the next upgrade with a fresh sequence.
+			upgradeSequence++
+
+			k.SetUpgradeErrorReceipt(ctx, portID, channelID, errorReceipt)
+			k.SetUpgradeSequence(ctx, portID, channelID, upgradeSequence)
+
+			// TODO: emit error receipt events
```

Comment:
I don't mind doing error receipt events in a follow up PR, considering this is already quite large